Make index Reducible

mpenet opened the pull request:
related to xtdb/xtdb#2034 

Baby steps, I didn't touch where it's used, just modified the impl. so that we can get there eventually.

* added `crux.index/idx->series` that returns a reified index "serie" that can be consumed as a reducible or a seq

* replace the impl of `crux.index/idx->seq` to leverage `crux.index/idx->series`

* updated test so that both new impl are verified as valid (seq version still works as expected and reducible works on similar test using into+xf)

* killed some un-used imports on modified namespaces

Still to do: 

* doing something similar for `crux.index/layered-idx->seq`
* building more high level api surface on top of new reducible interface

hraberg commented on src/crux/bootstrap.clj:
```diff
@@ -22,7 +21,7 @@
 
 (def default-options {:bootstrap-servers "localhost:9092"
-                      :group-id (.getHostName (InetAddress/getLocalHost))
+                      :group-id (.getHostName (InetAddress/getLoopbackAddress))
```
Why this change? Using localhost isn't optimal (these can be overridden via config), but the point of the group id is that it symbolises the current machine, using the loopback address would make all nodes be in the same group by default, which could be a good idea potentially, but then we could just hard code it to some default string "0" or so.

mpenet replied:
It was failing on my machine somehow, I didn't really dig and just changed it to loopback. But yeah I agree it can be improved.

hraberg replied:
Ah, interesting, what OS are you on? Will check with others and see if anyone else has issues with this. Could you give more information about what issue you ran into when not changing it?

Regardless, it can't be lookback before merging :-)

mpenet replied:
You're welcome! 

No numbers, it was my "lunch time hack" so I couldn't really go deep. I suspect the seq/fn call overhead will be negligible but I am fairly sure the reducible path will be more garbage friendly and faster if/when used appropriately (via into/transduce/educe etc).

mpenet replied:
> Ah, interesting, what OS are you on? Will check with others and see if anyone else has issues with this. Could you give more information about what issue you ran into when not changing it?

```
$ uname -a
Linux dixie 4.4.0-145-generic #171-Ubuntu SMP Tue Mar 26 12:43:40 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

the failure I mentioned: 

```shell

$ lein test
Retrieving ring-cors/ring-cors/0.1.13/ring-cors-0.1.13.pom from clojars
Retrieving ring-cors/ring-cors/0.1.13/ring-cors-0.1.13.jar from clojars
Compiling 2 source files to /home/mpenet/src/crux/target/classes
warning: [path] bad path element "/home/mpenet/src/crux/dev-resources": no such file or directory
warning: [path] bad path element "/home/mpenet/.m2/repository/org/roaringbitmap/RoaringBitmap/0.7.42/lib/shims-0.7.42.jar": no such file or directory
Exception in thread "main" Syntax error compiling at (bootstrap.clj:25:47).
	at clojure.lang.Compiler$StaticMethodExpr.eval(Compiler.java:1740)
	at clojure.lang.Compiler$InstanceMethodExpr.eval(Compiler.java:1545)
	at clojure.lang.Compiler$MapExpr.eval(Compiler.java:3054)
	at clojure.lang.Compiler$DefExpr.eval(Compiler.java:457)
	at clojure.lang.Compiler.eval(Compiler.java:7181)
	at clojure.lang.Compiler.load(Compiler.java:7635)
	at clojure.lang.RT.loadResourceScript(RT.java:381)
	at clojure.lang.RT.loadResourceScript(RT.java:372)
	at clojure.lang.RT.load(RT.java:463)
	at clojure.lang.RT.load(RT.java:428)
	at clojure.core$load$fn__6824.invoke(core.clj:6126)
	at clojure.core$load.invokeStatic(core.clj:6125)
	at clojure.core$load.doInvoke(core.clj:6109)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.core$load_one.invokeStatic(core.clj:5908)
	at clojure.core$load_one.invoke(core.clj:5903)
	at clojure.core$load_lib$fn__6765.invoke(core.clj:5948)
	at clojure.core$load_lib.invokeStatic(core.clj:5947)
	at clojure.core$load_lib.doInvoke(core.clj:5928)
	at clojure.lang.RestFn.applyTo(RestFn.java:142)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$load_libs.invokeStatic(core.clj:5985)
	at clojure.core$load_libs.doInvoke(core.clj:5969)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$require.invokeStatic(core.clj:6007)
	at clojure.core$require.doInvoke(core.clj:6007)
	at clojure.lang.RestFn.invoke(RestFn.java:1289)
	at crux.bootstrap.standalone$eval361$loading__6706__auto____362.invoke(standalone.clj:1)
	at crux.bootstrap.standalone$eval361.invokeStatic(standalone.clj:1)
	at crux.bootstrap.standalone$eval361.invoke(standalone.clj:1)
	at clojure.lang.Compiler.eval(Compiler.java:7176)
	at clojure.lang.Compiler.eval(Compiler.java:7165)
	at clojure.lang.Compiler.load(Compiler.java:7635)
	at clojure.lang.RT.loadResourceScript(RT.java:381)
	at clojure.lang.RT.loadResourceScript(RT.java:372)
	at clojure.lang.RT.load(RT.java:463)
	at clojure.lang.RT.load(RT.java:428)
	at clojure.core$load$fn__6824.invoke(core.clj:6126)
	at clojure.core$load.invokeStatic(core.clj:6125)
	at clojure.core$load.doInvoke(core.clj:6109)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.core$load_one.invokeStatic(core.clj:5908)
	at clojure.core$load_one.invoke(core.clj:5903)
	at clojure.core$load_lib$fn__6765.invoke(core.clj:5948)
	at clojure.core$load_lib.invokeStatic(core.clj:5947)
	at clojure.core$load_lib.doInvoke(core.clj:5928)
	at clojure.lang.RestFn.applyTo(RestFn.java:142)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$load_libs.invokeStatic(core.clj:5985)
	at clojure.core$load_libs.doInvoke(core.clj:5969)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$require.invokeStatic(core.clj:6007)
	at clojure.core$require.doInvoke(core.clj:6007)
	at clojure.lang.RestFn.invoke(RestFn.java:482)
	at crux.api_test$eval355$loading__6706__auto____356.invoke(api_test.clj:1)
	at crux.api_test$eval355.invokeStatic(api_test.clj:1)
	at crux.api_test$eval355.invoke(api_test.clj:1)
	at clojure.lang.Compiler.eval(Compiler.java:7176)
	at clojure.lang.Compiler.eval(Compiler.java:7165)
	at clojure.lang.Compiler.load(Compiler.java:7635)
	at clojure.lang.RT.loadResourceScript(RT.java:381)
	at clojure.lang.RT.loadResourceScript(RT.java:372)
	at clojure.lang.RT.load(RT.java:463)
	at clojure.lang.RT.load(RT.java:428)
	at clojure.core$load$fn__6824.invoke(core.clj:6126)
	at clojure.core$load.invokeStatic(core.clj:6125)
	at clojure.core$load.doInvoke(core.clj:6109)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.core$load_one.invokeStatic(core.clj:5908)
	at clojure.core$load_one.invoke(core.clj:5903)
	at clojure.core$load_lib$fn__6765.invoke(core.clj:5948)
	at clojure.core$load_lib.invokeStatic(core.clj:5947)
	at clojure.core$load_lib.doInvoke(core.clj:5928)
	at clojure.lang.RestFn.applyTo(RestFn.java:142)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$load_libs.invokeStatic(core.clj:5985)
	at clojure.core$load_libs.doInvoke(core.clj:5969)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$require.invokeStatic(core.clj:6007)
	at clojure.core$require.doInvoke(core.clj:6007)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:660)
	at user$eval237.invokeStatic(form-init1740504933893714412.clj:1)
	at user$eval237.invoke(form-init1740504933893714412.clj:1)
	at clojure.lang.Compiler.eval(Compiler.java:7176)
	at clojure.lang.Compiler.eval(Compiler.java:7166)
	at clojure.lang.Compiler.load(Compiler.java:7635)
	at clojure.lang.Compiler.loadFile(Compiler.java:7573)
	at clojure.main$load_script.invokeStatic(main.clj:452)
	at clojure.main$init_opt.invokeStatic(main.clj:454)
	at clojure.main$init_opt.invoke(main.clj:454)
	at clojure.main$initialize.invokeStatic(main.clj:485)
	at clojure.main$null_opt.invokeStatic(main.clj:519)
	at clojure.main$null_opt.invoke(main.clj:516)
	at clojure.main$main.invokeStatic(main.clj:598)
	at clojure.main$main.doInvoke(main.clj:561)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.lang.Var.applyTo(Var.java:705)
	at clojure.main.main(main.java:37)
Caused by: java.net.UnknownHostException: dixie: dixie: Name or service not known
	at java.net.InetAddress.getLocalHost(InetAddress.java:1506)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at clojure.lang.Reflector.invokeMatchingMethod(Reflector.java:167)
	at clojure.lang.Compiler$StaticMethodExpr.eval(Compiler.java:1733)
	... 102 more
Caused by: java.net.UnknownHostException: dixie: Name or service not known
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
	at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:929)
	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1324)
	at java.net.InetAddress.getLocalHost(InetAddress.java:1501)
	... 108 more
Tests failed.
```

hraberg replied:
Ah interesting, seems you're running into something like this:
https://stackoverflow.com/questions/7348711/recommended-way-to-get-hostname-in-java

We should arguably fix this, but don't necessarily need to do it as part of this PR. If you just back this out or now (potentially opening another issue about this), we should be good.

hraberg commented on src/crux/index.clj:
```diff
@@ -464,11 +462,29 @@
+(defn idx->series
+  [idx]
+  (reify
+    clojure.lang.IReduceInit
```
Do we need to implement both arities here - the one without init as well?

On another note, can we add imports to these `clojure.lang` things and remove the use of FQN inline.

mpenet replied:
No need: https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/IReduceInit.java

hraberg replied:
Ah, but is there a benefit or drawback of not implementing normal `IReduce` as well? Not saying we need it, just recall adding both at some point when doing something similar, but doesn't mean it was better!

mpenet replied:
I don't think so: 

> Collections or sequences can mark themselves as being internally reducible by implementing either IReduceInit or IReduce. Prior to 1.7, IReduce covered both forms of reduce (with and without the initial value). As of 1.7, these forms have been split apart - IReduceInit implements just the version with the initial value. IReduce now extends IReduceInit and adds the no-init version.

related: 
http://insideclojure.org/2014/12/21/reducible-repeat/
http://insideclojure.org/2015/01/18/reducible-generators/

hraberg replied:
Cool, we can always add add it if needed.

mpenet replied:
another ref about this: https://groups.google.com/forum/#!topic/clojure-dev/SNGLEJ4POnw

I personally never had to implement IReduce, but your mileage may vary. 

others seem to do it that way as well: 
https://github.com/seancorfield/next-jdbc/blob/788015909f7bbfb1ca54e465c65b9ffdf77764df/src/next/jdbc/result_set.clj

hraberg commented on src/crux/index.clj:
```diff
@@ -464,11 +462,29 @@
+              @ret
+              (recur ret)))
+          ret)))
+    clojure.lang.Seqable
```
Same as above. Add import.

hraberg commented on src/crux/index.clj:
```diff
@@ -464,11 +462,29 @@
+        (->> (repeatedly #(db/next-values idx))
+             (take-while identity)
+             (cons result))))
+    clojure.lang.Sequential))
```
Same as above. Add import.

hraberg reviewed (changes requested):
Thanks a lot, was wondering if this was still happening or not.

Looks good as far as I can see. Let's see if we can convert `layered-idx->seq` as well.
I assume you have no numbers yet? Would be interesting to see some comparison before and after, but can leave that until we have the second part.

Again, thanks a lot.

mpenet committed:
Add crux.index/idx->series returning reducible/seq "serie". #147

- Returns an index "serie" that can be consumed as Reducible or Seq

- Replace the impl of crux.index/idx->seq to leverage
  crux.index/idx->series

- Update test so that both new impl are verified as valid (seq version
  still works as expected and reducible works on similar test using
  into+xf)
